### DISCUSSION TITLE
Flexible cable

### DISCUSSION THREAD

**Discussion by DebbieDebbieli** [google-deepmind/mujoco#1001] [General]
I  want to simulate the motion of the cable wrapped on robot arm, likes https://www.youtube.com/watch?v=Ni7OSXxPASA
1) There are some fixed points on the cable which is used to attach it to arms.
2) The cable may be twisted when the robot arm rotating.
Can plugin cable in mujoco simulate this scenario？Any more tutorial or example for this?

**Comment by quagla** [comment]
Hi @DebbieDebbieli, the cable plugin can do what you want. I would suggest starting from https://github.com/deepmind/mujoco/blob/main/model/plugin/cable.xml

Load the model then select `Save xml` from the GUI and open it with a text editor. You will see that there is one capsule per segment of the cable, so you can easily add constraint to attach it to the robot arms. The `composite` object is just a keyword for not having to write a long XML, but if you load and save it as I suggested, you get an equivalent version that allows you to customize it.

**Reply by DebbieDebbieli** [reply]
Thanks for your guide. I add the robot arm(iiwa14) in the cable.xml .To attach the cable to the arm, I add the following constraint:
```
<equality>
      <connect body1="B_first" body2="link2" anchor="0 0 0">
      <connect body1="B_last" body2="link7" anchor="0 0 0">
</equality>
```
But both robot and cable move strangely. Something wrong with the constraints?
[kuka-iiwa-14-cable.zip](https://github.com/deepmind/mujoco/files/12227148/kuka-iiwa-14-cable.zip)

**Reply by quagla** [reply]
It looks like the cable starts from an unfeasible position, i.e. it crosses right through the arm?

**Reply by DebbieDebbieli** [reply]
Yes, I set an invalid position for cable. After changing the "offset" and "vertex" of the cable, it can move with the robot. There are severals problem:
1. The cable may be segmented when the distance of two vertice is large. In the saved xml I find the number of body in the cable is related with the number with vertex. The body/joint/geom in cable is only created at the "vertex"? 
2. Sometimes the cable may pass through the robot arm. Can I set a safe distance to avoid this contact?
![screenshot](https://github.com/deepmind/mujoco/assets/16498562/83966405-f620-4251-a141-fe86650c59a8)

**Reply by quagla** [reply]
1. There is one geom created in between each pair of vertices. You can increase the number of vertices but it will make the simulation slower. There is really no way to get around the fact that soft bodies in reality have an infinite number of degrees of freedom (hence the name "finite" elements in their approximation).
2. I think the cable is penetrating because the contact is too soft. You could make the cable slightly longer or try with 

```
  <default>
    <geom solref="0.01 1" solimp=".95 .99 .0001"/>
  </default>
```
